src/Interface/Modules/Render/ViewSceneControlsDock.cc

```diff
@@ -227,6 +234,74 @@ ViewSceneControlsDock::ViewSceneControlsDock(const QString& name, ViewSceneDialo
   transparencyGroupBox_->setVisible(false);
 }
 
+static bool vsdPairComp(std::pair<ViewSceneDialog*, bool> a, std::pair<ViewSceneDialog*, bool> b)
+{
+  return std::get<0>(a)->getName() < std::get<0>(b)->getName();
+}
+
+void ViewSceneControlsDock::updateViewSceneTree()
+{
+  viewSceneTreeWidget_->clear();
+
+  int numGroups = ViewSceneDialog::viewSceneManager.getGroupCount();
+  std::vector<ViewSceneDialog*> ungroupedMemebers;
+  ViewSceneDialog::viewSceneManager.getUngroupedViewScenesAsVector(ungroupedMemebers);
+
+  for(int i = 0; i < numGroups; ++i)
+  {
+    auto group = new QTreeWidgetItem(viewSceneTreeWidget_, QStringList(QString::fromStdString("Group:" + std::to_string(i))));
+    viewSceneTreeWidget_->addTopLevelItem(group);
+    group->setData(1, Qt::EditRole, i);
+
+    std::vector<ViewSceneDialog*> groupMemebers;
+    ViewSceneDialog::viewSceneManager.getViewSceneGroupAsVector(i, groupMemebers);
+
+    std::vector<std::pair<ViewSceneDialog*, bool>> viewScenesToDisplay;
+    for(ViewSceneDialog* vsd : groupMemebers) viewScenesToDisplay.push_back(std::make_pair(vsd, true));
+    for(ViewSceneDialog* vsd : ungroupedMemebers) viewScenesToDisplay.push_back(std::make_pair(vsd, false));
+    std::sort(viewScenesToDisplay.begin(), viewScenesToDisplay.end(), vsdPairComp);
+
+    for(int j = 0; j < viewScenesToDisplay.size(); ++j)
+    {
+      auto item = new QTreeWidgetItem(group, QStringList(QString::fromStdString(std::get<0>(viewScenesToDisplay[j])->getName())));
+      item->setCheckState(0, (std::get<1>(viewScenesToDisplay[j])) ? Qt::Checked : Qt::Unchecked);
+      item->setData(1, Qt::EditRole, (qulonglong)std::get<0>(viewScenesToDisplay[j]));
+    }
+  }
+
+  viewSceneTreeWidget_->expandAll();
+}
+
+void ViewSceneControlsDock::addGroup()
+{
+  ViewSceneDialog::viewSceneManager.addGroup();
+  groupRemoveSpinBox_->setRange(0, ViewSceneDialog::viewSceneManager.getGroupCount() - 1);
+}
+
+void ViewSceneControlsDock::removeGroup()
+{
+  uint32_t group = groupRemoveSpinBox_->value();
+  ViewSceneDialog::viewSceneManager.removeGroup(group);
+  groupRemoveSpinBox_->setRange(0, ViewSceneDialog::viewSceneManager.getGroupCount() - 1);
+}
+
+void ViewSceneControlsDock::viewSceneTreeClicked(QTreeWidgetItem* widgetItem, int column)
+{
+  QTreeWidgetItem* p = widgetItem->parent();
+  if(!p) return;
+  uint32_t g = p->data(1, Qt::EditRole).toInt();
+  if (widgetItem->checkState(column) == Qt::Unchecked)
+  {
+    ViewSceneDialog* vs = (ViewSceneDialog*)widgetItem->data(1, Qt::EditRole).toULongLong();
```

Comment:
If you must store pointers on the tree, you can use `QVariant::fromValue` and `value`, which are templated and can convert in between pointer types safely.

Reply:
Yeah, that's a lot cleaner I totally missed that when I read the QVarient documentation.